src/handlers/testing/exec/util/ci.ts

```diff
@@ -0,0 +1,176 @@
+import { z } from 'zod';
+import { getExecOutput } from '@actions/exec';
+import fs from 'fs/promises';
+
+export interface CIContext {
+  provider: 'github';
+  buildHtmlUrl: string;
+  repoId: string;
+  repoName: string;
+  repoHtmlUrl: string;
+  branchName: string;
+  defaultBranchName: string;
+  commitSha: string;
+  commitHtmlUrl: string;
+  commitMessage: string;
+  commitCommitterName: string;
+  commitCommitterEmail: string;
+  commitAuthorName: string;
+  commitAuthorEmail: string;
+  commitCommittedDate: string;
+  pullRequestNumber: number | null;
+  pullRequestTitle: string | null;
+  pullRequestHtmlUrl: string | null;
+}
+
+const zGitHubEnvSchema = z.object({
+  GITHUB_ACTIONS: z.literal('true'),
+  GITHUB_SERVER_URL: z.string(),
+  GITHUB_REPOSITORY: z.string(),
+  GITHUB_REF_NAME: z.string(),
+  GITHUB_EVENT_PATH: z.string(),
+  GITHUB_SHA: z.string(),
+  GITHUB_RUN_ID: z.string(),
+  GITHUB_RUN_ATTEMPT: z.string(),
+});
+
+const zCommitSchema = z.object({
+  sha: z.string(),
+  message: z.string(),
+  authorName: z.string(),
+  authorEmail: z.string(),
+  committerName: z.string(),
+  committerEmail: z.string(),
+  committedDate: z.string(),
+});
+
+const zPullRequestSchema = z.object({
+  number: z.number(),
+  title: z.string(),
+  head: z.object({
+    ref: z.string(),
+  }),
+  html_url: z.string(),
+});
+
+const zRepositorySchema = z.object({
+  id: z.coerce.string(),
+  default_branch: z.string(),
+});
+
+type Commit = z.infer<typeof zCommitSchema>;
+type PullRequest = z.infer<typeof zPullRequestSchema>;
+type Repository = z.infer<typeof zRepositorySchema>;
+
+function pullRequestFromEvent(event: unknown): PullRequest | null {
+  const parsed = zPullRequestSchema.safeParse(event);
+  if (parsed.success) {
+    return parsed.data;
+  }
+  return null;
+}
```

Comment:
should we log the error here?

Reply:
no because failures are expected for events that aren't pull request events (like `push` and `schedule`)